dev/com.ibm.ws.logging.core/src/com/ibm/ws/collector/manager/buffer/BufferManagerImpl.java

```diff
@@ -16,278 +16,278 @@
-            if (retSeqNum > seqNum) {
-                long eventsLost = retSeqNum - seqNum;
-                lostEventsForWarning += eventsLost;
-                lostEventsForTrace += eventsLost;
-                totalLostEvents += eventsLost;
-            }
-            long currentTime = System.currentTimeMillis();
-            long timeElapsed = currentTime - lastReportTimeForWarning;
-            if (lostEventsForWarning > 0 && timeElapsed >= intervalForWarning) {
-                if (lastReportTimeForWarning == 0) {
-                    // This was the first instance where we started to see loss
-                    // of events.
-                    Tr.warning(tc, "HANDLER_STARTED_TO_LOSE_EVENTS_WARNING", handlerId, lostEventsForWarning, sourceId);
-                } else {
-                    long timeElapsedInMins = TimeUnit.MILLISECONDS.toMinutes(timeElapsed);
-                    Tr.warning(tc, "HANDLER_LOST_EVENTS_WARNING", handlerId, lostEventsForWarning, sourceId,
-                            timeElapsedInMins, totalLostEvents);
-                }
-                lastReportTimeForWarning = currentTime;
-                lostEventsForWarning = 0;
-            }
-            timeElapsed = currentTime - lastReportTimeForTrace;
-            if (timeElapsed >= intervalForTrace) {
-                if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
-                    long timeElapsedInSecs = TimeUnit.MILLISECONDS.toSeconds(timeElapsed);
-                    Tr.event(tc,
-                            "Handler [{0}] has lost {1} events from source [{2}] in the last {3} second(s), and has lost {4} events from the source since the handler started.",
-                            handlerId, lostEventsForTrace, sourceId, timeElapsedInSecs, totalLostEvents);
-                }
-                lastReportTimeForTrace = currentTime;
-                lostEventsForTrace = 0;
-            }
-        }
-
-    }
+	private static final TraceComponent tc = Tr.register(BufferManagerImpl.class);
+	private Buffer<Object> ringBuffer;
+	private Set<SynchronousHandler> synchronousHandlerSet = new HashSet<SynchronousHandler>();
+
+	private final int capacity;
+
+	private final String sourceId;
+	/* Map to keep track of the next event for a handler */
+	private final ConcurrentHashMap<String, HandlerStats> handlerEventMap = new ConcurrentHashMap<String, HandlerStats>();
+
+	protected Queue<Object> earlyMessageQueue;
+
+	private static final int EARLY_MESSAGE_QUEUE_SIZE = 400;
+
+	public BufferManagerImpl(int capacity, String sourceId) {
+		super();
+
+		BufferManagerEMQHelper.addBufferManagerList(this);
+		ringBuffer = null;
+		this.sourceId = sourceId;
+		this.capacity = capacity;
+		if (!BufferManagerEMQHelper.getEMQRemovedFlag()) {
+			earlyMessageQueue = new SimpleRotatingSoftQueue<Object>(new Object[EARLY_MESSAGE_QUEUE_SIZE]);
+			//Check again just in case
+			if (BufferManagerEMQHelper.getEMQRemovedFlag()){
```

Comment:
getEMQRemovedFlag needs to be synchronized, otherwise change of emqRemovedFlag value may not copy to other threads before it is read